crates/apollo-compiler/src/diagnostic.rs

```diff
@@ -0,0 +1,174 @@
+//! Pretty-printable diagnostic reports for custom errors that reference GraphQL documents.
+use crate::validation::FileId;
+use crate::validation::NodeLocation;
+use crate::SourceFile;
+use crate::SourceMap;
+use ariadne::ColorGenerator;
+use ariadne::ReportKind;
+use std::fmt;
+use std::io;
+use std::ops::Range;
+use std::sync::Arc;
+use std::sync::OnceLock;
+
+type MappedSpan = (FileId, Range<usize>);
+
+/// Translate a byte-offset location into a char-offset location for use with ariadne.
+fn map_span(sources: &SourceMap, location: NodeLocation) -> Option<MappedSpan> {
```

Comment:
I’ve added these public APIs in https://github.com/apollographql/apollo-rs/pull/758 but they may belong more in this PR (or separately):
```rust
impl GraphQLLocation {
    /// Convert a `NodeLocation` to a line and column number
    pub fn from_node(sources: &SourceMap, location: Option<NodeLocation>) -> Option<Self> {…}
}
impl Node {
    pub fn line_column(&self, sources: &SourceMap) -> Option<GraphQLLocation> {…}
}
impl NodeStr {
    pub fn line_column(&self, sources: &SourceMap) -> Option<GraphQLLocation> {…}
}
```